src/scheduled_callbacks.toit

```diff
@@ -0,0 +1,139 @@
+// Copyright (C) 2024 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the LICENSE file.
+
+import monitor
+
+class PriorityQueueElement:
+  value/any
+  priority/int
+  position_/int := -1
+
+  constructor .value --.priority:
+
+  is-linked -> bool:
+    return position_ != -1
+
+/**
+A simple priority queue implementation.
+
+Uses a heap to store the elements.
+Each element knows its position in the heap, so that we can update the heap in O(log n) time.
+*/
+class PriorityQueue:
+  heap_/List ::= []
+
+  size -> int:
+    return heap_.size
+
+  add element/PriorityQueueElement -> none:
+    heap_.add element
+    element.position_ = heap_.size - 1
+    bubble-up_ (heap_.size - 1)
+
+  remove element/PriorityQueueElement -> none:
+    index := element.position_
+    // Swap the element with the last element.
+    heap_[index] = heap_.last
+    heap_.resize (heap_.size - 1)
+    // Move the element down the tree.
+    bubble-down_ index
+    element.position_ = -1
+
+  first -> PriorityQueueElement:
+    return heap_.first
+
+  bubble-up_ index/int:
+    heap := heap_
+    while index > 0:
+      parent-index := (index - 1) / 2
+      current := heap[index]
+      parent := heap[parent-index]
+      if current.priority < parent.priority:
+        // Swap the elements.
+        heap[index] = parent
+        parent.position_ = index
+        heap[parent-index] = current
+        current.position_ = parent-index
+        // Move up the tree.
+        index = parent-index
+      else:
+        break
+
+  bubble-down_ index/int:
+    heap := heap_
+    if index >= heap.size: return
+    while true:
+      current := heap[index]
+      left-index := 2 * index + 1
+      right-index := 2 * index + 2
+      if left-index >= heap.size:
+        break
+      smallest-index := left-index
+      if right-index < heap.size and heap[right-index].priority < heap[left-index].priority:
+        smallest-index = right-index
+      if current.priority < heap[smallest-index].priority:
+        break
+      // Swap the elements.
+      heap[index] = heap[smallest-index]
+      heap[index].position_ = index
+      heap[smallest-index] = current
+      current.position_ = smallest-index
+      // Move down the tree.
+      index = smallest-index
+
+class ScheduledCallbacks:
```

Comment:
I think this is a bit of a weird name. I think this could be just a `Schedule` with a `add` and `remove` methods.